### PR TITLE
ci: LSDV-4734: Setup Cypress integration tests

### PR DESCRIPTION
**Pull request by nicholasrq**
### Description
This branch is to introduce Cypress as the main frontend framework for Functional and Integration testing.
We're migrating from Codecept to Cypress due to instability and flakiness of the first. Cypress will provide more robust and structured tests that can run faster and safer.

### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [x] Tests for the changes have been added/updated (for bug fixes/features)
- [x] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
- [x] CI
- [x] Testing

### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [x] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [x] integration
- [ ] unit

### CODE REVIEW

File: .github/workflows/cicd_pipeline.yml [hunk 1]
```diff
@@ -46,6 +46,15 @@ jobs:
     with:
       sha: ${{ github.event.pull_request.head.sha || github.event.after }}
 
+  run_functional:
+    name: "Tests"
+    if: github.event_name == 'push' || github.event.pull_request.draft == false
+    uses: heartexlabs/label-studio-frontend/.github/workflows/fun_tests.yml@fb-lsdv-4734/cypress-init
```
Comment (farioas) [inline .github/workflows/cicd_pipeline.yml:52]: Don't forget to change branch to `master` before merge

File: .github/workflows/fun_tests.yml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  workflow_call:
+    inputs:
+      sha:
+        required: true
+        type: string
+
+env:
+  # increment it in case if you need to reset cache
+  CACHE_NAME_PREFIX: v3
+  NODE: '14'
```
Comment (farioas) [inline .github/workflows/fun_tests.yml:13]: node 16?

Comment (nicholasrq) [reply .github/workflows/fun_tests.yml:13]: that's a separate task, we have to update to node LTS version everywhere, but currently nothing works on 16+

File: .github/workflows/cicd_pipeline.yml (hunk 1 above)
Comment (nicholasrq) [reply .github/workflows/cicd_pipeline.yml:52]: sure!

File: __tests__/cypress.config.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import configure from '@heartexlabs/ls-test';
```
Comment (Gondragos) [inline __tests__/cypress.config.js:1]: Right now there is a problem to run it on Window 10. 
https://github.com/heartexlabs/ls-frontend-test/blob/main/configure.js#L11
`process.env.PWD` is `undefined`

File: public/index.html [hunk 4]
```diff
@@ -87,7 +87,12 @@
...
       ]
       domReady(function () {
-        var ls = new LabelStudio("label-studio", {
+        if (window.DEFAULT_LSF_INIT === false) {
+          console.log("Default initialization prevented")
+          return false;
+        }
```
Comment (Gondragos) [inline public/index.html:93]: Thank you!

File: __tests__/cypress.config.js (hunk 3 above)
Comment (nicholasrq) [reply __tests__/cypress.config.js:1]: thx, will take a look into it. as a workaround for now you can try running it under WSL. unfortunately, I don't have windows right now to check it

Comment (nicholasrq) [reply __tests__/cypress.config.js:1]: oh, it looks like we can replace it with `process.cwd()` instead

File: public/index.html (hunk 4 above)
Comment (nicholasrq) [reply public/index.html:93]: haha, yeah, default init was causing weird behavior. now tests can init their own LSF without blinking screen and different artifacts left by the previous instance

File: __tests__/cypress.config.js (hunk 3 above)
Comment (hlomzik) [inline __tests__/cypress.config.js:1]: Is it a good name for this dir? `__tests__` is usually used for tests in a context of some logical domain and here we have full cypress setup. Maybe `cy`, `cypress` or `tests/cypress`?

File: __tests__/specs/core/feature_flags.cy.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { LabelStudio } from '@heartexlabs/ls-test/helpers/LSF';
+
+describe('Feature Flags', () => {
+  it('can set feature flags on the global object', () => {
+    const flagName = 'customFeatureFlag';
+
+    cy.visit('/');
+
+    LabelStudio.setFeatureFlags({
+      [flagName]: true,
```
Comment (hlomzik) [inline __tests__/specs/core/feature_flags.cy.ts:10]: I'd check two flags with different values to be sure that this is not a default value in use.
```suggestion
      [enabledFlag]: true,
      [disabledFlag]: false,
```

File: src/utils/feature-flags.ts [hunk 6]
```diff
@@ -212,6 +212,16 @@ export const FF_LSDV_4701 = 'fflag_feat_front_lsdv_4701_audio_default_decoder_ff
  */
 export const FF_LSDV_4659 = 'fflag_feat_front_lsdv_4659_skipduplicates_060323_short';
 
+Object.assign(window, {
+  APP_SETTINGS: {
+    ...(window.APP_SETTINGS ?? {}),
+    feature_flags: {
+      ...(window.APP_SETTINGS?.feature_flags ?? {}),
+      ...(window.__FEATURE_FLAGS__ ?? {}),
```
Comment (hlomzik) [inline src/utils/feature-flags.ts:220]: Why can't we just change `APP_SETTINGS` without additional global var?

File: public/index.html [hunk 7]
```diff
@@ -142,7 +147,11 @@
             }
           },
           history: annotationHistory,
-        });
+        };
+
+        Object.assign(lsfConfig, window.LSF_CONFIG ?? {});
```
Comment (hlomzik) [inline public/index.html:152]: I think you can just use `assignConfig()` method

**Review by hlomzik — CHANGES_REQUESTED** [review]
Please, fill in the PR description and other fields

File: __tests__/cypress.config.js (hunk 3 above)
Comment (Gondragos) [reply __tests__/cypress.config.js:1]: In my case works
```JavaScript
const localPath = p => path.resolve(process.cwd(), p);
```
but I don't know if it is cross platform.

File: __tests__/specs/core/feature_flags.cy.ts (hunk 5 above)
Comment (nicholasrq) [reply __tests__/specs/core/feature_flags.cy.ts]: fair point. added your suggestion + added better helper for testing this stuff

File: __tests__/cypress.config.js (hunk 3 above)
Comment (nicholasrq) [reply __tests__/cypress.config.js]: makes sense. moved to `tests/functional`

File: __tests__/cypress/support/e2e.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { CURRENT_FLAGS } from '../../feature-flags';
+import '@heartexlabs/ls-test/cypress/support/e2e';
+
+beforeEach(() => {
+  cy.on('window:before:load', (win) => {
+    console.log('Setting feature flags', CURRENT_FLAGS);
+    Object.assign(win, {
+      DEFULT_LSF_INIT: false,
```
[the remark this answers is not in the document]
Comment (nicholasrq) [reply __tests__/cypress/support/e2e.ts]: fixed. didn't work as expected. thanks for noticing

File: public/index.html (hunk 7 above)
Comment (nicholasrq) [reply public/index.html:152]: turned out we don't need it at all

File: __tests__/feature-flags.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  const featureFlags = getFeatureFlags();
+
+  if (id in featureFlags) {
+    return featureFlags[id] === true;
+  } else {
+    return window.APP_SETTINGS?.feature_flags_default_value === true;
+  }
+}
+
+Object.assign(window, { getFeatureFlags, isFF });
```
[the remark this answers is not in the document]
Comment (nicholasrq) [reply __tests__/feature-flags.ts]: true. removed

File: src/utils/feature-flags.ts (hunk 6 above)
Comment (nicholasrq) [reply src/utils/feature-flags.ts:220]: true. probably don't need it at all

**Review by Gondragos — APPROVED** [review]

File: src/utils/feature-flags.ts (hunk 6 above)
Comment (nicholasrq) [reply src/utils/feature-flags.ts:220]: turned out that without it feature flags that are set before page is loaded are overriden somehow. I will keep it in place for safety

**Review by hlomzik — APPROVED** [review]

File: tests/functional/package.json [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  "scripts": {
+    "test": "cypress run --quiet --browser chrome",
+    "test:ui": "cypress open",
+    "test:with-coverage": "COLLECT_COVERAGE=true yarn test:parallel",
+    "test:parallel": "cypress-parallel -r spec -b -m -s test -t ${CPU_NUMBER:=2} -d ./specs/",
+    "test:reset": "rm -rf output && yarn run cvg:reset",
+    "cvg:reset": "rm -rf coverage .nyc_output",
+    "cvg:merge": "rm -rf coverage && nyc merge ./.nyc_output ./coverage/coverage.json --cwd=. --exclude-after-remap false",
+    "cvg:report": "nyc report --temp-dir=.nyc_output --report-dir=coverage --reporter=clover --reporter=text --reporter=json --cwd=. --exclude-after-remap false",
+    "cvg:summary": "nyc report --temp-dir=.nyc_output --reporter=text-summary --cwd=. --exclude-after-remap false"
```
[the remark this answers is not in the document]
Comment (nicholasrq) [reply tests/functional/package.json:18]: for some reason commands with `coverage:` executed automatically after each test which is unwanted. that's why i shortened them

**Pull request merged**